app/gui2/src/components/GraphEditor/widgets/WidgetVector.vue

```diff
@@ -41,6 +42,62 @@ const value = computed({
 })
 
 const navigator = injectGraphNavigator(true)
+
+function useChildEditForwarding(widgetId: string) {
+  const editHandler = WidgetEditHandler.New(widgetId, props.input, {
+    cancel: () => (itemEdits.value = new Map()),
+    end: () => (itemEdits.value = new Map()),
+  })
+
+  const itemEdits = shallowRef(new Map<Ast.AstId, Ast.Owned | string>())
+  const updateEdits = (f: (itemEdits: Map<Ast.AstId, Ast.Owned | string>) => void) => {
+    f(itemEdits.value)
+    triggerRef(itemEdits)
+  }
+
+  function childEditHandler(widgetInput: WidgetInput, astId: Ast.AstId) {
+    return editHandler.child(`${widgetId}.Child`, widgetInput, {
+      cancel: () => updateEdits((itemEdits) => itemEdits.delete(astId)),
+      end: () => updateEdits((itemEdits) => itemEdits.delete(astId)),
+      edit: (_origin, value) => {
+        updateEdits((itemEdits) => itemEdits.set(astId, value))
+        return false
+      },
+    })
+  }
+
+  let editStarted = false
+  watchEffect(() => {
+    const inputValue = props.input.value
+    if (!(inputValue instanceof Ast.Ast)) return
+    if (!editStarted && itemEdits.value.size === 0) return
+    const editedAst = Ast.copyIntoNewModule(inputValue)
+    const module = editedAst.module
+    for (const [id, replacement] of itemEdits.value) {
+      const ast = module.tryGet(id)
+      if (!ast) {
+        continue
+      }
+      ast.replace(typeof replacement === 'string' ? Ast.parse(replacement, module) : replacement)
+    }
+    editHandler.edit(editedAst)
+    editStarted = true
+  })
+
+  return { childEditHandler }
+}
+
+const { childEditHandler } = useChildEditForwarding('WidgetVector')
+
+function itemInput(ast: Ast.Ast): WidgetInput {
+  const baseInput = WidgetInput.FromAst(ast)
+  return {
+    ...baseInput,
+    dynamicConfig: itemConfig.value,
+    forcePort: true,
+    editHandler: childEditHandler(baseInput, ast.id),
```

Comment:
So far, it was the _parent's_ edit handler which was passed in WidgetInput, so the constructor in WidgetHandler takes it and treats as a parent. I think it was better, because it was simpler; We just had one handler for each widget. Now we have handler of parent widget, then handlers for parent-widget relations, and finally the handler of the child itself (because the child might want to handle some interactions in its own way: `WidgetText` for example creates new WidgetHandler for this reason)

In my initial implementation, all hooks (including `edit` hook) presented portId as origin, to handle cases exactly like this (I was thinking about DateTime widget then). It was lost somewhere in the process, and sorry I didn't catch it on review. Let's restore it now.